File: scala/parallel_collection/download_web/download_web.scala
import scala.io.Source

import java.io.{File, InputStream}
import java.net.URL
import java.nio.file.{Paths, Files}
import java.nio.file.StandardCopyOption._

val dir = args(0)

val using = (st: InputStream) => (block: InputStream => Unit) => try {block(st)} finally {st.close()}

Source.stdin.getLines.toList.par.foreach {u =>
	val url = new URL(u)
	val f = new File(url.getFile()).getName()
	val filePath = Paths.get(dir, f)

	try {
		using (url.openStream()) {stream =>
			Files.copy(stream, filePath, REPLACE_EXISTING)
		}

		printf("completed: %s => %s\n", url, filePath)

	} catch {
		case e: Exception => printf("failed: %s, %s\n", url, e)
	}
}

File: java/7.0/nio2/download_web/DownloadWeb.java

import java.io.*;
import java.net.URI;
import java.util.concurrent.Executors;
import java.util.concurrent.ExecutorService;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;

public class DownloadWeb {
	public static void main(String[] args) throws Exception {
		if (args.length < 1) {
			System.out.println("java DownloadWeb <output dir>");
			return;
		}

		BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));

		ExecutorService exec = Executors.newCachedThreadPool();

		final String dir = args[0];
		String url = null;

		while ((url = reader.readLine()) != null) {

			final URI uri = URI.create(url);
			String fileName = new File(uri.getPath()).getName();
			final Path filePath = Paths.get(dir, fileName);

			exec.submit(new Runnable() {
				@Override
				public void run() {
					try (InputStream in = uri.toURL().openStream()) {
						Files.copy(in, filePath, StandardCopyOption.REPLACE_EXISTING);
						System.out.printf("completed: %s => %s\n", uri, filePath);
					} catch (Exception ex) {
						System.out.printf("failed: %s\n", uri);
					}
				}
			});
		}

		exec.shutdown();
	}
}